pkg/agent/multicast/mcast_controller.go

```diff
@@ -465,6 +476,105 @@ func (c *Controller) addOrUpdateGroupEvent(e *mcastGroupEvent) {
 	}
 }
 
+func (c *Controller) memberChanged(e interface{}) {
+	podEvent := e.(types.PodUpdate)
+	namespace, name := podEvent.PodNamespace, podEvent.PodName
+	containerID := podEvent.ContainerID
+	iface, ok := c.ifaceStore.GetContainerInterface(containerID)
+
+	klog.V(2).InfoS("Calling memberChanged", "IsAdd", podEvent.IsAdd, "namespace", namespace,
+		"name", name)
+	if podEvent.IsAdd {
+		if ok {
+			g := &mcastGroupEvent{
+				eType: groupJoin,
+				iface: iface,
+			}
+			c.queryGroupCh <- g
+		}
+	} else {
+		interfaceName := util.GenerateContainerInterfaceName(name, namespace, containerID)
+		g := &mcastGroupEvent{
+			eType: groupLeave,
+			iface: &interfacestore.InterfaceConfig{
+				InterfaceName: interfaceName,
+			},
+		}
+		c.queryGroupCh <- g
+		c.removeLocalInterface(interfaceName)
+	}
+}
+
+func (c *Controller) syncQueryGroup(stopCh <-chan struct{}) {
+	for {
+		klog.InfoS("Resync query group")
+		if c.queryGroupId == 0 {
+			klog.V(2).InfoS("As c.queryGroupId is 0, query group will not be installed")
+			return
+		}
+		select {
+		case event := <-c.queryGroupCh:
+			if event.eType == groupJoin {
+				if event.iface != nil {
+					c.queryGroupMember[event.iface.InterfaceName] = event.iface.OFPort
+					err := c.updateQueryGroup("")
+					if err != nil {
+						klog.ErrorS(err, "Failed to update query group for groupJoin")
+					}
+				}
+			} else if event.eType == groupLeave {
+				err := c.updateQueryGroup(event.iface.InterfaceName)
+				if err != nil {
+					klog.ErrorS(err, "Failed to update query group for groupLeave")
+				} else {
+					delete(c.queryGroupMember, event.iface.InterfaceName)
+				}
+			}
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+func (c *Controller) updateQueryGroup(staleIface string) error {
+	groupKey := types.McastAllHosts.String()
+	memberPorts := make([]uint32, 0)
+	for ifaceName, port := range c.queryGroupMember {
```

Comment:
What do you think if we use a set for ofport as this function's parameter, and get the slice in the caller?